# Homework 2 (Due: 2021.04.01 11:59 PM)

In [ ]:
Name:ju won ung

Student ID:2015312490

For this homework, you are provided with an input file.
Submit the code that you used for analysis as an ipython notebook and an exported PDF file.

# Q1 Analyzing the Election Data (3 x 8 = 24 pts)

Run the following cell to set the filename.

In [1]:
import pandas as pd
filename = 'bush-gore-results-fl_demo.csv'

(a) Determine how many counties Bush won. You can assume that each row corresponds to a unique county.
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [2]:
#
# YOUR CODE HERE
#
df = pd.read_csv(filename)
length = df.shape[0]

df_1 = df.loc[:, ['bush', 'gore']]
"""bush_won = 0
for i in range(0, length):
    if df_1.loc[i, 'bush'] > df_1.loc[i, 'gore']:
        bush_won += 1"""
        
df_bush_won = df_1.loc[df_1['bush'] > df_1['gore']]
print(df_bush_won)
print("count of bush won %d" % df_bush_won.shape[0])


      bush    gore
1     5710    2492
2    38737   18950
3     5514    3175
4   115285   97418
6     2973    2255
7    35526   29745
8    29865   25625
9    41836   14732
10   60533   30018
11   11064    7147
12    4356    3420
13    2797    1926
14  152198  107964
15   73117   41043
17    2554    2146
19    3400    2010
20    1941    1542
21    3650    2497
22    2246    1822
23    3865    2439
24    4847    3340
26   20306   14267
27  180860  169657
28    5111    2277
29   28735   19868
30    9238    6968
32    1770     889
33   50110   36671
34  106241   73660
36    6958    5498
37    1417    1117
38    3138    3114
39   58052   49277
40   55241   44765
41   34070   26720
44   16380    6979
45   52193   17048
46    5157    4688
52   90280   75293
53   13547   12202
54   36374   12902
55   83200   72953
56   75777   59274
57   39646   19602
59   12227    9737
60    8106    4175
61    4156    2749
62    2432    1507
64    4612    3938
65   12282    5742
66    5094    2898
count of bus

In [3]:
df

co   lat   lon    npop  whit  blac  hisp   o65  hsed  coll  ...    bush  \
0    1  31.7  84.4  198326  74.4  21.8   4.7   9.4  82.7  34.6  ...   34224   
1    2  32.3  84.3   20761  82.4  16.8   1.5   7.7  64.1   5.7  ...    5710   
2    3  32.2  87.6  146223  84.2  12.4   2.4  11.9  74.7  15.7  ...   38737   
3    4  31.9  84.2   24646  76.1  22.9   2.6  11.8  65.0   8.1  ...    5514   
4    5  30.3  82.7  460977  88.3   9.2   4.1  16.5  82.3  20.4  ...  115285   
..  ..   ...   ...     ...   ...   ...   ...   ...   ...   ...  ...     ...   
62  63  32.0  84.4   12359  71.0  27.8   4.8   7.0  67.7   7.9  ...    2432   
63  64  31.1  83.1  419797  88.0  10.5   5.0  22.7  75.4  14.8  ...   82314   
64  65  32.2  86.4   19172  83.9  14.9   0.9  10.9  71.6  10.9  ...    4612   
65  66  32.6  88.2   37914  88.9   8.6   1.2  14.9  66.5  11.9  ...   12282   
66  67  32.6  87.7   20221  79.7  17.6   1.5  16.4  60.9   7.4  ...    5094   

     gore  brow  nade  harr  hage  buch  mcre  phil  moor  
0   47465   758  3326   106   142   363   104   120   121  
1    2492   117   153   100   103   173   100   103   103  
2   18950   271   928   105   118   348   103   118   127  
3    3175   128   184   100   102   165   100   102   103  
4   97418   743  4570   111   139   670   111   172   176  
..    ...   ...   ...   ...   ...   ...   ...   ...   ...  
62   1507   115   133   101   100   137   100   101   100  
63  97163   542  3003   108   136   596   105   120   169  
64   3938   130   249   102   103   146   101   100   106  
65   5742   168   365   103   111   220   102   107   118  
66   2898   132   193   100   102   188   100   109   105  

[67 rows x 21 columns]

(b) Determine who won the largest county. The column named "npop" records the size of each county.
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [4]:
#
# YOUR CODE HERE
#
npop_idx = df.loc[:, 'npop'].idxmax(axis=0)

df_npop = df.loc[npop_idx, 'bush':]
candidate1 = df_npop.idxmax(axis=0)
print(candidate1)

gore


(c) Determine the average number of votes per county that Buchanan obtained. You can assume that each row corresponds to a unique county.
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [5]:
#
# YOUR CODE HERE
#
df_buch = df.loc[:, ['buch']]
buch_mean = df.buch.mean()
print(buch_mean)


360.67164179104475


(d) Determine, for the number of votes Buchanan obtained in Palm Beach, how many standard deviations it is away from Buchanan's overall mean, in absolute value. The row with coounty number 50 ("co"=50) records the results for Palm Beach County. (You can assumed that such a row will exist in the test case.)
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [6]:
#
# YOUR CODE HERE
#
df_buch = df.loc[:, ['buch']]
buch_mean = float(df_buch.mean())
buch_std = float(df_buch.std())
x = float(df.loc[df['co']==50, 'buch'])

"""df_palm = df.loc[df['co']==50]
palm_std = df_palm.loc[:, 'bush':]"""

print(abs(float((x-buch_mean)/buch_std)))


6.993018491215867


(e) Now calculate the above statistic (same as in part f) for all the counties and report them in decreasing order.
    Demo file: 'bush-gore-results-fl_demo.csv' 
    Example output: 
    county_50 6.993018  
    county_52 ...
    ...
    ...

In [7]:
#
# YOUR CODE HERE
#
"""df_buch = df.loc[:, ['buch']]
buch_mean = float(df_buch.mean())
buch_std = float(df_buch.std())"""

list = []

for idx, value in df_buch.iterrows():
    list.append(abs(float((value-buch_mean)/buch_std)))
    
desc_or = pd.DataFrame({"co":range(1, df.shape[0]+1)})
desc_or.loc[:, 'value'] = pd.Series(list)
desc_or_sorted = desc_or.sort_values(by=['value'], axis=0, ascending=False)

for idx, value in desc_or_sorted.iterrows():
    print("county_%-2d %.6f\n" % (int(value.iloc[0]), value.iloc[1]))


county_50 6.993018

county_52 1.672122

county_28 1.303171

county_6  1.172038

county_15 0.869765

county_51 0.687512

county_5  0.687512

county_41 0.671954

county_43 0.665286

county_53 0.603053

county_21 0.559365

county_33 0.557142

county_16 0.536376

county_25 0.530471

county_23 0.528248

county_64 0.523040

county_62 0.519358

county_20 0.514913

county_39 0.514913

county_32 0.514913

county_14 0.514913

county_24 0.512690

county_18 0.506022

county_13 0.499354

county_63 0.497132

county_19 0.494909

county_38 0.492687

county_47 0.483796

county_65 0.477128

county_44 0.474906

county_4  0.434899

county_37 0.430454

county_22 0.421564

county_2  0.417118

county_48 0.411910

county_29 0.410451

county_17 0.394892

county_67 0.383779

county_12 0.381557

county_45 0.379334

county_7  0.379334

county_31 0.352663

county_30 0.345995

county_61 0.339327

county_42 0.330437

county_60 0.325992

county_66 0.312656

county_11 0.308211

county_59 0.303766

county_27 0.297098



(f) Assuming that the votes were distributed across the white, black, and hispanic population uniformly, determine which candidate obtained the largest number of votes for each subpopulation.
    Demo file: 'bush-gore-results-fl_demo.csv' 
    Example output:
    white: bush
    black: ..
    hispanic: ..

In [8]:
#
# YOUR CODE HERE
#
candidate_list = ['bush', 'gore', 'brow', 'nade', 'harr', 'hage', 'buch', 'mcre', 'phil', 'moor']
df_f = df.loc[:, ['npop', 'whit', 'blac', 'hisp', 'bush', 'gore', 'brow', 'nade', 'harr', 'hage', 'buch', 'mcre', 'phil', 'moor']]
df_temp = df.loc[:, ['whit', 'blac', 'hisp']]
df_per = df_temp.sum(axis=1)
df_f['per'] = df_per
#make npop uniformly
"""
whit_n = []
blac_n = []
hisp_n = []

for idx, value in df_f.iterrows():
    whit_n.append(value.loc['npop']*value.loc['whit']/100.0)
    blac_n.append(value.loc['npop']*value.loc['blac']/100.0)
    hisp_n.append(value.loc['npop']*value.loc['hisp']/100.0)
#find the percentage of votes of candidates
df_per = df_f.loc[:, 'bush':].sum(axis=1)
df_f['total'] = df_per"""

length = df_f.shape[0]
whit_per = []
blac_per = []
hisp_per = []
for idx, row in df_f.iterrows():
    temp = []
    for name in candidate_list:
        temp.append(row.loc[name]*row.loc['whit']/row.loc['per'])
    whit_per.append(temp)
    
    temp = []
    for name in candidate_list:
        temp.append(row.loc[name]*row.loc['blac']/row.loc['per'])
    blac_per.append(temp)
    
    temp = []
    for name in candidate_list:
        temp.append(row.loc[name]*row.loc['hisp']/row.loc['per'])
    hisp_per.append(temp)


    
df_whit = pd.DataFrame(whit_per, columns = candidate_list)
df_blac = pd.DataFrame(blac_per, columns = candidate_list)
df_hisp = pd.DataFrame(hisp_per, columns = candidate_list)

df_whit_sum = df_whit.sum(axis=0)
df_blac_sum = df_blac.sum(axis=0)
df_hisp_sum = df_hisp.sum(axis=0)

print('the candidate obtainged the largest number of votes for white people : %s' %(df_whit_sum.idxmax()))
print('the candidate obtainged the largest number of votes for black people : %s' %(df_blac_sum.idxmax()))
print('the candidate obtainged the largest number of votes for hispanic people : %s' %(df_hisp_sum.idxmax()))

#print(df_whit[==df_whit.sum(axis=0).max())

the candidate obtainged the largest number of votes for white people : bush
the candidate obtainged the largest number of votes for black people : gore
the candidate obtainged the largest number of votes for hispanic people : gore


(g) Calculate the correlation between the difference in votes between Bush and Gore, and the votes obtained by Nader.
(FYI: Pearson's correlation coefficient) https://en.wikipedia.org/wiki/Pearson_correlation_coefficient
    Demo file: 'bush-gore-results-fl_demo.csv' 

In [9]:
#
# YOUR CODE HERE
#
import math
df_corr = df.loc[:,['bush','gore', 'nade']]
#
df_bush = df.loc[:, 'bush']
df_gore = df.loc[:, 'gore']
df_nade = df.loc[:, 'nade']
#diff bush and gore
df_bg = abs(df_bush-df_gore)

#power2 
pow_nade = df_nade*df_nade
pow_bg = df_bg*df_bg

corr_bgn = ((df_bg*df_nade).mean() - df_bg.mean()*df_nade.mean()) / (math.sqrt( pow_bg.mean() - math.pow(df_bg.mean(), 2))*math.sqrt( pow_nade.mean() - math.pow(df_nade.mean(), 2)))
print('correlation between the difference in votes between Gore and Nader: %.6f\n' %corr_bgn)


df_bg = pd.DataFrame(abs(df_bush-df_gore), columns=['diff'])
df_bgn = pd.concat([df_bg, df_nade], axis=1)
df_bgn.corr(method='pearson')




correlation between the difference in votes between Gore and Nader: 0.544306



diff      nade
diff  1.000000  0.544306
nade  0.544306  1.000000

(h) Find the distance between the county that Bush won by the largest margin and the county that Gore won by the largest margin. (Just use basic Euclidean distance between the latitude (lat) and longitude (lon) values for the counties, no need to compute spherical distance.) (FYI: Eucliden distane is decribed in https://en.wikipedia.org/wiki/Euclidean_distance#:~:text=In%20mathematics%2C%20the%20Euclidean%20distance,metric%20as%20the%20Pythagorean%20metric.)
    Demo file: 'bush-gore-results-fl_demo.csv' 
   

In [10]:
#
# YOUR CODE HERE
#
import math
df_h = df.loc[:, ['bush', 'gore']]

df_margin = df_f.loc[:, 'bush'] - df_f.loc[:, 'gore']
bw = df_margin.idxmax()
gw = df_margin.idxmin()
x1, y1 = df.loc[bw, 'lat':'lon']
x2, y2 = df.loc[gw, 'lat':'lon']
print(x1, y1)
print(x2, y2)
euc_dist = math.sqrt(math.pow(x1-x2, 2) + math.pow(y1-y2, 2))
print('Euclidean Distance : %.6f\n' % euc_dist)


32.3 83.6
28.1 82.5
Euclidean Distance : 4.341659

